src/client.rs

```diff
@@ -1,928 +1,594 @@
-    ReadOnly,
-    ModifyDisabled,
-    ServiceFailed,
-    Timeout,
-    Unknown,
-}
-
-impl From<u32> for WriteSettingError {
-    fn from(n: u32) -> Self {
-        #[cfg(target_os = "windows")]
-        let n = n as i32;
-
-        #[allow(non_upper_case_globals)]
-        match n {
-            settings_write_res_e_SETTINGS_WR_VALUE_REJECTED => WriteSettingError::ValueRejected,
-            settings_write_res_e_SETTINGS_WR_SETTING_REJECTED => WriteSettingError::SettingRejected,
-            settings_write_res_e_SETTINGS_WR_PARSE_FAILED => WriteSettingError::ParseFailed,
-            settings_write_res_e_SETTINGS_WR_READ_ONLY => WriteSettingError::ReadOnly,
-            settings_write_res_e_SETTINGS_WR_MODIFY_DISABLED => WriteSettingError::ModifyDisabled,
-            settings_write_res_e_SETTINGS_WR_SERVICE_FAILED => WriteSettingError::ServiceFailed,
-            settings_write_res_e_SETTINGS_WR_TIMEOUT => WriteSettingError::Timeout,
-            _ => WriteSettingError::Unknown,
+        let fields = split_multipart(&msg.setting);
+        if let [group, name, value] = fields.as_slice() {
+            let setting = Setting::new(group, name);
+            let value = SettingValue::parse(value, setting.kind);
+            Ok(Entry { setting, value })
+        } else {
+            Err(Error::ParseError)
         }
     }
 }
 
-impl From<i32> for WriteSettingError {
-    fn from(n: i32) -> Self {
-        (n as u32).into()
-    }
-}
+impl TryFrom<MsgSettingsReadResp> for Entry {
+    type Error = Error;
 
-impl std::fmt::Display for WriteSettingError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            WriteSettingError::ValueRejected => {
-                write!(f, "setting value invalid",)
-            }
-            WriteSettingError::SettingRejected => {
-                write!(f, "setting does not exist")
+    fn try_from(msg: MsgSettingsReadResp) -> Result<Self, Self::Error> {
+        let fields = split_multipart(&msg.setting);
+        match fields.as_slice() {
+            [group, name] => {
+                let setting = Setting::new(&group, &name);
+                Ok(Entry {
+                    setting,
+                    value: None,
+                })
             }
-            WriteSettingError::ParseFailed => {
-                write!(f, "could not parse setting value ")
-            }
-            WriteSettingError::ReadOnly => {
-                write!(f, "setting is read only")
-            }
-            WriteSettingError::ModifyDisabled => {
-                write!(f, "setting is not modifiable")
-            }
-            WriteSettingError::ServiceFailed => write!(f, "system failure during setting"),
-            WriteSettingError::Timeout => write!(f, "request wasn't replied in time"),
-            WriteSettingError::Unknown => {
-                write!(f, "unknown settings write response")
+            [group, name, value] | [group, name, value, ..] => {
```

Comment:
Why would there be more than group, name, value out of interest?

Reply:
Good point I don't that should happen, I'll try out removing it